### PR TITLE
Epic6.5 notify invites - mk [WIP]

### PR DESCRIPTION
**Pull request by mattkrick**
Underlying business logic:
- [x] review the spec for handling invitations here: https://github.com/ParabolInc/action/blob/epic6.5-notify-invites/src/server/graphql/mutations/helpers/inviteTeamMembers/__tests__/getAction.test.js

User Settings > Notifications
- [x] notification components are loaded asynchronously (test by putting a console.log in one just below the imports & see that it does not log until that notification is used)

Logged in as a billing leader >> Team Settings
- [x] Invite someone already on the team, see an error
- [x] Invite someone already invited, see an error
- [x] invite someone with pending approval on the same team, see an error.
- [x] invite someone with pending approval on a different team in the same org. see an INVITED toast. see the name with an invited status below. 
browser 2 (inviter): get an INVITEE_APPROVED notification
- [x] invite someone with pending invitation on a different team in the same org. see an INVITED toast. see the name with an invited status below. see the invitation on the other team is not affected. 
- [x] Invite someone who is not on parabol, see that an invitation is emailed to them & it shows invitation sent in the team settings list below. see an invite toast.
- [x] invite someone who is on parabol, but not in the org. see an invite toast. 
browser 2 (invitee): see that they receive an invite via notification
server console: see an email invitation. 
- [x] invite someone who is on parabol & in the org, but has never been on the team. see an invite toast.
browser 2 (invitee) see that they receive an invite notification.
server console: see an email invitation.
- [x] invite someone who is active on parabol & in the org & used to be on the team. see a reactivation toast. 
browser 2 (invitee): see they are immediately added to the team, they receive a notification telling them they're back on the team. the new team is visible in their sidebar.
browser 3 (another team member): see a toast showing who rejoined the team. 
- [x] cancel an invite for someone who was a part of another org.
browser 2 (invitee): see the teamInvite notification go away
- [x] kick a team member off the team
browser 2 (kicked off person): see a toast saying you've been removed. get an updated auth token. ~~see the team disappear from the side bar~~ (limited by cashay, will test later). get a notification making you acknowledge it. no longer subscribed to that channel (have a 3rd browser rejoin the team & make sure a toast doesn't appear)
 
Logged in as a non-billing leader >> Team Settings
- [x] invite the billing leader who has never been on the team. see they are immediately invited below. see an invite toast. 
browser 2 (billing leader): see a new invite notification.
- [x] invite the billing leader who used to be on the team. see they are immediately reactivated. see a reactivation toast. 
browser 2 (billing leader): see a new notification. 
browser 3 (another team member): see a rejoin toast mentioning the billing leader.
- [x] invite someone not in the org. see that it turns into an org approval. 
browser 2 (billing leader): see a notification come through requesting approval.
- [x] clicking resend invitation will pop a toast & show a new invited at date
- [x] grabbing the new invitation url from the server console after clicking "resend", the url will allow the person to join the team 
- [x] clicking cancel invitation will remove the invite
- [x] invite someone who was invited by the billing leader on another team. see that the invitation goes through

Notifications >> REQUEST_NEW_USER >> Deny
- [x] as a billing leader, see that you have a request to add a new person to the org. 
- [x] attempt to deny w/o a reason. the field should show an error saying "required"
- [x] deny with a reason & see that the notification goes away.
browser 2 (another billing leader on the same org): see that the notification went away in real time when the first billing leader denied it
tab 2: see the notification goes away, too.
- [x] refresh the page & see that the notification is still gone

Notifications >> REQUEST_NEW_USER >> Accept
- [x] as a billing leader, click accept to the new user. See the notification goes away.
browser 2 (inviter): see a new INVITEE_APPROVED notification telling them that the billing leader has approved it. see the status update in team settings list.
browser 3 (invitee): see a new TEAM_INVITE notification
server console: see an email go to the invitee
 
Notifications >> INVITEE_APPROVED
- [x] as a non-billing leader that invited someone new to the team, after the billing leader approved the invitee, see a notification come through saying that person was approved.
- [x] when you acknowledge that notification, it goes away

Notifications >> TEAM_INVITE

Welcome Wizard >> Ste 2
- [x] Can create a team

Welcome Wizard >> Step 3
- [x] Invite 1 person on parabol & 1 brand new person. see 2 emails go out & 1 notification added. 

New Team >> New Org
- [x] Invite 1 person currently on parabol but different org, 1 currently in the same org, and 1 new to parabol. See toasts pop up showing they were invited.
server console: see 3 emails go out

New Team >> New Team
- [x]  Perform the above, see the same results.

Meeting
- [x] ask to become the facilitator
browser 2 (facilitator) get a toast asking to promote a guy to facilitator

Organizations >> Members
- [x] Promote a member to billing leader
browser 2 (the promotee): see a toast & a notification that you've been promoted

### CODE REVIEW

**Comment by mattkrick** [comment]
Currently, we hard delete OrgApproval documents from the database.
This is a liability because a customer could argue that they didn't approve X to join the org.
Also, this makes it a headache if someone was approved to join team A, and invitation to team A was sent, and then team B asked to invite the same person. We'd then have to look over all invitations to that org. It'd be so much nicer to have a single table showing all approval requests & who accepted/rejected each.
If team A & B seek approval, it'd created 2 rows. Then, when either was approved, it'd approve both atomically.

**Comment by mattkrick** [comment]
This thing is HUGE, so here's the big picture:
- the invitations module was super duper buggy. Everything from sending 2x invite emails, to broken "resend invitation" functionality, there were some glaring errors we had to git rid of before going public beta.
- There were a ton of edge cases not accounted for with invites. For example, let's say i invite someone new, then Jordan (the billing leader) invites him. My invite is still stuck at pending approval, but his invite was sent! Him inviting my guy should also approve of my guy. Similarly, let's say Jordan invites someone on his private team & then i try to invite that someone on a team that Jordan isn't on. It should be auto approved because 1 invite already went out from the org. But what if Jordan cancels his invite? Well if i already had an invite sent out, it stands. but if it hasn't gone out, then I'll need approval again.
- To solve the above, and also the occasional customer complaining "i didn't approve X", we don't remove records from OrgApprovals anymore. Instead, an approval starts as active & pending & is either approved or denied. If the person gets kicked off the org, then it goes to inactive.
- Whether it's a Welcome wizard, a new org, a new team, or a single invite from team settings, all invites follow the same flow & have the same predictable notifications. 
- Notifications grew & changed as the app changed, which means the logic hung out in the socketcluster middleware, the dashboard container, and the socket container. Now, it all travels the same path. Instead of using the socketcluster exchange, a notification gets pushed to redis & then goes down to individual users or teams. if the message contains an authToken, we update it on the socket server, _not within graphql_. This opens to door to us breaking graphql out into its own stateless microservice. On the client, each notification is handled on a per-type basis. Some only pop toasts, some pop toasts & leave permanent messages. The one exception is Presence-- i ran out of time to refactor that so we'll save it for the dashboard redesign.
- Notification details used to be stored in an array to work with cashay & a bug with GraphQLInterfaceType. This means you needed a legend to read the data. Now each notification type is a superset of a general notification & it's declarative & beautiful.
- Notification types are loaded async. That means if we have 100 different notification types, we don't have to load 100 of those components on startup, just the ones that user is looking at.

**Comment by jordanh** [comment]
Hero's work sir. I'll start tests now. (Family dinner went a bit longer than I expected this eve!)

**Comment by jordanh** [comment]
# Test Cycle Template
## Status: in-progress/compeled

## Tests

### User Settings > Notifications
- [ ] notification components are loaded asynchronously (test by putting a console.log in one just below the imports & see that it does not log until that notification is used)

### Logged in as a billing leader >> Team Settings
- [ ] Invite someone already on the team, see an error
- [ ] Invite someone already invited, see an error
- [ ] invite someone with pending approval on the same team, see an error.
- [ ] invite someone with pending approval on a different team in the same org. see an INVITED toast. see the name with an invited status below. 
browser 2 (inviter): get an INVITEE_APPROVED notification
- [ ] invite someone with pending invitation on a different team in the same org. see an INVITED toast. see the name with an invited status below. see the invitation on the other team is not affected. 
- [ ] Invite someone who is not on parabol, see that an invitation is emailed to them & it shows invitation sent in the team settings list below. see an invite toast.
- [ ] invite someone who is on parabol, but not in the org. see an invite toast. 
browser 2 (invitee): see that they receive an invite via notification
server console: see an email invitation. 
- [ ] invite someone who is on parabol & in the org, but has never been on the team. see an invite toast.
browser 2 (invitee) see that they receive an invite notification.
server console: see an email invitation.
- [ ] invite someone who is active on parabol & in the org & used to be on the team. see a reactivation toast. 
browser 2 (invitee): see they are immediately added to the team, they receive a notification telling them they're back on the team. the new team is visible in their sidebar.
browser 3 (another team member): see a toast showing who rejoined the team. 
- [ ] cancel an invite for someone who was a part of another org.
browser 2 (invitee): see the teamInvite notification go away
- [ ] kick a team member off the team
browser 2 (kicked off person): see a toast saying you've been removed. get an updated auth token. ~~see the team disappear from the side bar~~ (limited by cashay, will test later). get a notification making you acknowledge it. no longer subscribed to that channel (have a 3rd browser rejoin the team & make sure a toast doesn't appear)
 
### Logged in as a non-billing leader >> Team Settings
- [ ] invite the billing leader who has never been on the team. see they are immediately invited below. see an invite toast. 
browser 2 (billing leader): see a new invite notification.
- [ ] invite the billing leader who used to be on the team. see they are immediately reactivated. see a reactivation toast. 
browser 2 (billing leader): see a new notification. 
browser 3 (another team member): see a rejoin toast mentioning the billing leader.
- [ ] invite someone not in the org. see that it turns into an org approval. 
browser 2 (billing leader): see a notification come through requesting approval.
- [ ] clicking resend invitation will pop a toast & show a new invited at date
- [ ] grabbing the new invitation url from the server console after clicking "resend", the url will allow the person to join the team 
- [ ] clicking cancel invitation will remove the invite
- [ ] invite someone who was invited by the billing leader on another team. see that the invitation goes through

### Notifications >> REQUEST_NEW_USER >> Deny
- [ ] as a billing leader, see that you have a request to add a new person to the org. 
- [ ] attempt to deny w/o a reason. the field should show an error saying "required"
- [ ] deny with a reason & see that the notification goes away.
browser 2 (another billing leader on the same org): see that the notification went away in real time when the first billing leader denied it
tab 2: see the notification goes away, too.
- [ ] refresh the page & see that the notification is still gone

### Notifications >> REQUEST_NEW_USER >> Accept
- [ ] as a billing leader, click accept to the new user. See the notification goes away.
browser 2 (inviter): see a new INVITEE_APPROVED notification telling them that the billing leader has approved it. see the status update in team settings list.
browser 3 (invitee): see a new TEAM_INVITE notification
server console: see an email go to the invitee
 
### Notifications >> INVITEE_APPROVED
- [ ] as a non-billing leader that invited someone new to the team, after the billing leader approved the invitee, see a notification come through saying that person was approved.
- [ ] when you acknowledge that notification, it goes away

### Notifications >> TEAM_INVITE

### Welcome Wizard >> Step 2
- [ ] Can create a team

### Welcome Wizard >> Step 3
- [ ] Invite 1 person on parabol & 1 brand new person. see 2 emails go out & 1 notification added. 

### New Team >> New Org
- [ ] Invite 1 person currently on parabol but different org, 1 currently in the same org, and 1 new to parabol. See toasts pop up showing they were invited.
server console: see 3 emails go out

### New Team >> New Team
- [ ]  Perform the above, see the same results.

### Meeting
- [ ] ask to become the facilitator
browser 2 (facilitator) get a toast asking to promote a guy to facilitator

### Organizations >> Members
- [ ] Promote a member to billing leader
browser 2 (the promotee): see a toast & a notification that you've been promoted

**Comment by jordanh** [comment]
# Test Cycle 1
## Status: in-progress

## Tests

### User Settings > Notifications
- [ ] notification components are loaded asynchronously (test by putting a console.log in one just below the imports & see that it does not log until that notification is used)

### Logged in as a billing leader >> Team Settings
- [ ] Invite someone already on the team, see an error
- [ ] Invite someone already invited, see an error
- [ ] invite someone with pending approval on the same team, see an error.
- [ ] invite someone with pending approval on a different team in the same org. see an INVITED toast. see the name with an invited status below. 
browser 2 (inviter): get an INVITEE_APPROVED notification
- [ ] invite someone with pending invitation on a different team in the same org. see an INVITED toast. see the name with an invited status below. see the invitation on the other team is not affected. 
- [ ] Invite someone who is not on parabol, see that an invitation is emailed to them & it shows invitation sent in the team settings list below. see an invite toast.
- [ ] invite someone who is on parabol, but not in the org. see an invite toast. 
browser 2 (invitee): see that they receive an invite via notification
server console: see an email invitation. 
- [ ] invite someone who is on parabol & in the org, but has never been on the team. see an invite toast.
browser 2 (invitee) see that they receive an invite notification.
server console: see an email invitation.
- [ ] invite someone who is active on parabol & in the org & used to be on the team. see a reactivation toast. 
browser 2 (invitee): see they are immediately added to the team, they receive a notification telling them they're back on the team. the new team is visible in their sidebar.
browser 3 (another team member): see a toast showing who rejoined the team. 
- [ ] cancel an invite for someone who was a part of another org.
browser 2 (invitee): see the teamInvite notification go away
- [ ] kick a team member off the team
browser 2 (kicked off person): see a toast saying you've been removed. get an updated auth token. ~~see the team disappear from the side bar~~ (limited by cashay, will test later). get a notification making you acknowledge it. no longer subscribed to that channel (have a 3rd browser rejoin the team & make sure a toast doesn't appear)
 
### Logged in as a non-billing leader >> Team Settings
- [ ] invite the billing leader who has never been on the team. see they are immediately invited below. see an invite toast. 
browser 2 (billing leader): see a new invite notification.
- [ ] invite the billing leader who used to be on the team. see they are immediately reactivated. see a reactivation toast. 
browser 2 (billing leader): see a new notification. 
browser 3 (another team member): see a rejoin toast mentioning the billing leader.
- [ ] invite someone not in the org. see that it turns into an org approval. 
browser 2 (billing leader): see a notification come through requesting approval.
- [ ] clicking resend invitation will pop a toast & show a new invited at date
- [ ] grabbing the new invitation url from the server console after clicking "resend", the url will allow the person to join the team 
- [ ] clicking cancel invitation will remove the invite
- [ ] invite someone who was invited by the billing leader on another team. see that the invitation goes through

### Notifications >> REQUEST_NEW_USER >> Deny
- [ ] as a billing leader, see that you have a request to add a new person to the org. 
- [ ] attempt to deny w/o a reason. the field should show an error saying "required"
- [ ] deny with a reason & see that the notification goes away.
browser 2 (another billing leader on the same org): see that the notification went away in real time when the first billing leader denied it
tab 2: see the notification goes away, too.
- [ ] refresh the page & see that the notification is still gone

### Notifications >> REQUEST_NEW_USER >> Accept
- [ ] as a billing leader, click accept to the new user. See the notification goes away.
browser 2 (inviter): see a new INVITEE_APPROVED notification telling them that the billing leader has approved it. see the status update in team settings list.
browser 3 (invitee): see a new TEAM_INVITE notification
server console: see an email go to the invitee
 
### Notifications >> INVITEE_APPROVED
- [ ] as a non-billing leader that invited someone new to the team, after the billing leader approved the invitee, see a notification come through saying that person was approved.
- [ ] when you acknowledge that notification, it goes away

### Notifications >> TEAM_INVITE

### Welcome Wizard >> Step 2
- [ ] Can create a team

### Welcome Wizard >> Step 3
- [ ] Invite 1 person on parabol & 1 brand new person. see 2 emails go out & 1 notification added. 

### New Team >> New Org
- [ ] Invite 1 person currently on parabol but different org, 1 currently in the same org, and 1 new to parabol. See toasts pop up showing they were invited.
server console: see 3 emails go out

### New Team >> New Team
- [ ]  Perform the above, see the same results.

### Meeting
- [ ] ask to become the facilitator
browser 2 (facilitator) get a toast asking to promote a guy to facilitator

### Organizations >> Members
- [ ] Promote a member to billing leader
browser 2 (the promotee): see a toast & a notification that you've been promoted

**Comment by jordanh** [comment]
Ok, I found a couple of bugs and had a few questions. All in all this is looking _great_. Hero's work sir! 🏆

File: src/client/Atmosphere.js [hunk 1]
```diff
@@ -86,7 +86,7 @@ export default class Atmosphere extends Environment {
         variables
       })
     });
-    const resJson = res.json();
+    const resJson = await res.json();
```
Comment (jordanh) [inline src/client/Atmosphere.js]: oooh nice

File: src/server/graphql/types/Notification.js [hunk 2]
```diff
@@ -0,0 +1,91 @@
+import {GraphQLID, GraphQLInterfaceType, GraphQLList, GraphQLNonNull} from 'graphql';
```
Comment (jordanh) [inline src/server/graphql/types/Notification.js]: This is a dope pattern

**Comment by mattkrick** [comment]
> When removing a user from an organization

Fixed!

> Error: notification not found!

Can't reproduce! any extra details about this? You click the notification, not the toast, correct?

> When a billing leader receives the "requested approval / check it out" toast, it dismisses itself awfully quickly. We might want to bump up the time on that a bit. In fact, all the toasts dismiss themselves pretty quickly. Maybe I'm just too tired and not reading very fast 

we can slow it down if you like, personally i think it's OK, but then again i wrote the toasts, so my opinion doesn't count.

> I see when a person is kicked off a team that it doesn't yet disspear from the sidebar (as expected). We tackle that in the New UI epic?

Correct, it's that old cashay bug about referencing cached data because we can't dispose of it. We'll fix during the new UI epic

> When I promote a user to be billing leader and click the toast on the second browser, the
"Check It out!" button takes me to the Org settings, but the integrations tab. I take it this is because we don't have a way to link into a specific tab?

You say integrations, but i think you mean "Billing". I updated it so it goes to the "Members" tab.

**Comment by jordanh** [comment]
# Test Cycle 2
## Status: in-progress

## Tests

**Comment by jordanh** [comment]
> we can slow it down if you like, personally i think it's OK, but then again i wrote the toasts, so my opinion doesn't count.

Mind slowing it down like 50%? (make the delay twice as long?)

> You say integrations, but i think you mean "Billing". I updated it so it goes to the "Members" tab.

You're right. Heh. I had an error on recalling what was written on the first tab. 😉

**Comment by mattkrick** [comment]
OK, upped it from 5 to 10 seconds.

**Comment by jordanh** [comment]
Well shoot, now I can't reproduce it either. It's likely some bizarre edge-case. For now I think this is safe to try. Any objections to merging this? Anything else you'd like me to try and test before we do?
